src/librbd/io/AioCompletion.cc

```diff
@@ -252,7 +254,16 @@ void AioCompletion::complete_external_callback() {
     while (ictx->external_callback_completions.pop(aio_comp)) {
       aio_comp->complete_cb(aio_comp->rbd_comp, aio_comp->complete_arg);
       aio_comp->complete_event_socket();
+
+      if (aio_comp == this) {
+        // delay release of original AioCompletion to ensure ImageCtx is not
+        // destructed
+        release_aio_comp = true;
+        continue;
+      }
+
       aio_comp->notify_callbacks_complete();
+      aio_comp->put();
```

Comment:
@dillaman Just thinking. Suppose `this` completion is completed by another thread, i.e. `notify_callbacks_complete()` is called and the `async_op` is finished. Which allows the image context to be closed (if the close happens to be just waiting for inflight ops), Then this thread when trying to de-reference `ictx` below may find it invalid? Or am I missing something?

Reply:
Hmm, potentially. I tweaked it to always keep the most recently dequeued AioCompletion valid.